File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/AddBattleMessageClosure.java
package rustleund.fightingfantasy.framework.closures.impl;

import org.w3c.dom.Element;

import rustleund.fightingfantasy.framework.base.GameState;
import rustleund.fightingfantasy.framework.base.BattleState.BattleMessagePosition;

public class AddBattleMessageClosure extends AbstractClosure {

	private String message;
	private BattleMessagePosition messagePosition;

	public AddBattleMessageClosure(Element element) {
		this.message = element.getAttribute("message");
		if (element.hasAttribute("messagePosition")) {
			this.messagePosition = BattleMessagePosition.valueOf(element.getAttribute("messagePosition"));
		} else {
			this.messagePosition = BattleMessagePosition.END;
		}
	}

	@Override
	public boolean execute(GameState gameState) {
		gameState.getBattleState().addAdditionalMessage(this.messagePosition, this.message);
		return true;
	}

}

File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/AddBattleMessageClosure.kt
package rustleund.fightingfantasy.framework.closures.impl

import org.w3c.dom.Element
import rustleund.fightingfantasy.framework.base.BattleState.BattleMessagePosition
import rustleund.fightingfantasy.framework.base.GameState
import rustleund.fightingfantasy.framework.base.asElementSequence
import rustleund.fightingfantasy.framework.base.optionalAttribute
import rustleund.fightingfantasy.framework.base.writeTag
import rustleund.fightingfantasy.framework.closures.Closure

class AddBattleMessageClosure(element: Element) : Closure {

    private val message = element.optionalAttribute("message") ?: ""
    private val messagePosition = element.optionalAttribute("messagePosition")
        ?.let { BattleMessagePosition.valueOf(it) } ?: BattleMessagePosition.END
    private val messageBody = writeTag(element.childNodes.asElementSequence().firstOrNull())

    override fun execute(gameState: GameState): Boolean {
        if (messagePosition == BattleMessagePosition.IMMEDIATE_END) {
            gameState.battleState?.immediatelyAppendBattleMessage(message + messageBody)
        } else {
            gameState.battleState?.addAdditionalMessage(messagePosition, message + messageBody)
        }
        return true
    }
}